drivers/platform/mbed/mbed_irq.h

```diff
@@ -0,0 +1,149 @@
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#ifndef MBED_IRQ_H
+#define MBED_IRQ_H
+
+// Platform support needs to be C-compatible to work with other drivers
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdbool.h>
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/**
+ * @enum irq_id mapped with 'irq_ctrl_id' from 'irq_init_param' structure in
+ *       'no-os/irq.h' file.
+ * @brief Interrupt IDs supported by the mbed irq driver.
+ * @note Every interrupt instance should have unique interrupt ID. For instance,
+ * there are 5 interrupt IDs for external/gpio interrupts. The IDs can be mapped
+ * randomly to any of the instances as Mbed layer abstracts most of the low level
+ * MCU details such as IRQn_type, IRQn_number, etc.
+ * Example: struct irq_init_param my_ext_irq_init_param = {
+ *				.irq_ctrl_id = EXTERNAL_INT_ID1
+ *			}
+ */
+enum irq_id {
+	/* External/GPIO interrupt IDs (Max possible instances of external IRQs are
+	 * possible/implemented in the mbed irq module) */
+	/** External interrupt ID1 */
+	EXTERNAL_INT_ID1,
+	/** External interrupt ID2 */
+	EXTERNAL_INT_ID2,
+	/** External interrupt ID3 */
+	EXTERNAL_INT_ID3,
+	/** External interrupt ID4 */
+	EXTERNAL_INT_ID4,
+	/** External interrupt ID5 */
+	EXTERNAL_INT_ID5,
+
+	/** UART Rx interrupt ID1 (Multiple instances of UART Rx IRQs are possible
+	 ** but only one implemented in mbed irq module) */
+	UART_RX_INT_ID1,
+
+	/** Ticker interrupt ID (Only one instance of Ticker IRQ possible) */
+	TICKER_INT_ID,
+
+	/* Number of available interrupts */
+	NB_INTERRUPTS
+};
+
+/*
+ * @enum External/GPIO IRQ events to be set into 'int_mode' parameter.
+ * Example: struct mbed_irq_init_param my_ext_mbed_irq_init_param = {
+ *				.int_mode = EXT_IRQ_RISE
+ *			}
+ * */
+enum ext_irq_event {
```

Comment:
In my opinion, this event should go together into the `struct mbed_callback` i mentioned above.
When you configure a GPIO interrupt, you may want to have two callbacks:
```
void rising_edge_irq(void) {
    // do something on rising edge
}

void falling_edge_irq(void) {
    // do something on falling edge
}
```
This should be configurable with `irq_register_callback(desc, irq_id, callback)`.
I think something like:
```
struct mbed_callback rising {
    .callback = rising_edge_irq,
    .event = MBED_RISING // invented macro name, look up the correct one in mbed libraries
}

struct mbed_callback falling {
    .callback = falling_edge_irq,
    .event = MBED_FALLING // invented macro name, look up the correct one in mbed libraries
}

irq_register_callback(desc, irq_id, rising)
irq_register_callback(desc, irq_id, falling)
```